### PR TITLE
Add topic per message handling

### PR DESCRIPTION
**Pull request by Gnucki**

### CODE REVIEW

**Comment by Gnucki** [comment]
#9

File: QueueInteropTransport.php [hunk 1]
```diff
@@ -103,11 +104,18 @@ public function receive(callable $handler): void
     public function send(Envelope $message): void
     {
         $psrContext = $this->contextManager->psrContext();
-        $destination = $this->getDestination();
-        $topic = $psrContext->createTopic($destination['topic']);
+
+        $topicItem = $message->get(Topic::class);
```
Comment (sroze) [inline QueueInteropTransport.php]: I'd prefer to move this logic to the `getDestination` method instead. Can it take the message as a nullable argument?

File: QueueInteropTransport.php [hunk 2]
```diff
@@ -103,11 +104,18 @@ public function receive(callable $handler): void
...
+        if (null !== $topicItem) {
```
Comment (sroze) [inline QueueInteropTransport.php]: You can also use a one-liner:
```php
if (null !== ($topicItem = $message->get(Topic::class))) {
```

File: EnvelopeItem/Topic.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+namespace Enqueue\MessengerAdapter\EnvelopeItem;
+
+use Symfony\Component\Messenger\EnvelopeItemInterface;
+
+/**
+ * @author Thomas Prelot <tprelot@gmail.com>
+ *
+ * @experimental in 4.1
+ */
+final class Topic implements EnvelopeItemInterface
```
Comment (sroze) [inline EnvelopeItem/Topic.php]: I'd actually name it more generic; `TransportConfiguration`? Which would contain a `topic` string field.

**Comment by sroze** [comment]
Also, watch out the failing PHP-CS rules and tests :)

**Comment by Gnucki** [comment]
Thanks for the review.
I'll try to make a new version for the beginning of next week.

**Comment by Gnucki** [comment]
Done!
Tell me if I can do something more.

**Comment by sroze** [comment]
Thank you @Gnucki.

File: EnvelopeItem/TransportConfiguration.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+ * @author Thomas Prelot <tprelot@gmail.com>
+ *
+ * @experimental in 4.1
+ */
+final class TransportConfiguration implements EnvelopeItemInterface
+{
+    /**
+     * @param string $topic
+     */
+    public function __construct(array $configuration)
```
Comment (ogizanagi) [inline EnvelopeItem/TransportConfiguration.php]: Wouldn't providing directly the `?string $topic = null` be better instead of an arbitrary array?

Comment (Gnucki) [reply EnvelopeItem/TransportConfiguration.php]: Did it for a better time scalability.
An option resolver could also be a possibility.

Comment (sroze) [reply EnvelopeItem/TransportConfiguration.php]: Yup, I believe that it's better for later: we won't have to break the constructor when adding multiple type of parameters.

Comment (ogizanagi) [reply EnvelopeItem/TransportConfiguration.php]: Adding arguments in a constructor is never an issue BC wise (removing/altering is), but fine to me :)

Comment (sroze) [reply EnvelopeItem/TransportConfiguration.php]: Maybe we (at least I) were looking far too much in advance: I saw many additions then too many arguments so the need to refactor it into an array 😄

Comment (Gnucki) [reply EnvelopeItem/TransportConfiguration.php]: I thought the same :)

**Pull request merged**